ext/cache/01_cache.js

```diff
@@ -0,0 +1,312 @@
+        ObjectPrototypeIsPrototypeOf(URLPrototype, request)
+      ) {
+        r = new Request(request);
+      }
+      return await core.opAsync("op_cache_delete", {
+        cacheId: this[_id],
+        requestUrl: r.url,
+      });
+    }
+
+    /** See https://w3c.github.io/ServiceWorker/#cache-matchall
+     *
+     * Note: the function is private as we don't want to expose
+     * this API to the public yet.
+     *
+     * The function will return an array of responses.
+     */
+    async #matchAll(request, _options) {
+      // Step 1.
+      let r = null;
+      // Step 2.
+      if (ObjectPrototypeIsPrototypeOf(RequestPrototype, request)) {
+        r = request;
+        if (request.method !== "GET") {
+          return [];
+        }
+      } else if (
+        typeof request === "string" ||
+        ObjectPrototypeIsPrototypeOf(URLPrototype, request)
+      ) {
+        r = new Request(request);
+      }
+
+      // Step 5.
+      const responses = [];
+      // Step 5.2
+      if (r === null) {
+        // Step 5.3
+        // Note: we have to return all responses in the cache when
+        // the request is null.
+        // We deviate from the spec here and return an empty array
+        // as we don't expose matchAll() API.
+        return responses;
+      } else {
+        // Remove the fragment from the request URL.
+        const url = new URL(r.url);
+        url.hash = "";
+        const innerRequest = toInnerRequest(r);
+        const matchResult = await core.opAsync(
+          "op_cache_match",
+          {
+            cacheId: this[_id],
+            requestUrl: url.toString(),
+            requestHeaders: innerRequest.headerList,
+          },
+        );
+        if (matchResult) {
+          const [meta, responseBodyRid] = matchResult;
+          let body = null;
+          if (responseBodyRid !== null) {
+            body = new ReadableStream({
+              type: "bytes",
+              async pull(controller) {
+                try {
+                  // This is the largest possible size for a single packet on a TLS
+                  // stream.
+                  const chunk = new Uint8Array(16 * 1024 + 256);
+                  const read = await core.read(responseBodyRid, chunk);
+                  if (read > 0) {
+                    // We read some data. Enqueue it onto the stream.
+                    controller.enqueue(chunk.subarray(0, read));
+                  } else {
+                    // We have reached the end of the body, so we close the stream.
+                    core.close(responseBodyRid);
+                    controller.close();
+                  }
+                } catch (err) {
+                  // There was an error while reading a chunk of the body, so we
+                  // error.
+                  controller.error(err);
+                  controller.close();
+                  core.close(responseBodyRid);
+                }
+              },
+            });
```

Comment:
This can be replaced with:
```suggestion
            body = readableStreamFromRid(responseBodyRid);
```

Reply:
Do you mean to move the code into a function? I couldn't find readableStreamFromRid() defined anywhere.